### ISSUE TITLE
Incorrect NCBI taxonomy ids for Homo sapiens in interactions.tsv.gz file

### ISSUE THREAD

**Issue by pwrose** [globalbioticinteractions/globalbioticinteractions#569]
I noticed that references to Homo sapiens are assigned an NCBI taxonomy id:741158, which is an extinct human species.

I should be changed to NCBI taxonomy id:9606.

**Comment by jhpoelen** [comment]
Hey @pwrose - thanks for sharing your observation that Homo sapiens was suspiciously linked to ```Homo sapiens subsp. 'Denisova'``` NCBI:txid741158 as well as ```Homo sapiens``` NCBI:txid9606 . Note that the interactions.tsv.gz contains a list of taxonIds that includes these and other related taxon identifiers relevant to the source/target taxa.

As close inspection it appears that the link was introduced via a taxon name resolver (see previously reported issue https://github.com/GlobalNamesArchitecture/gni/issues/48 fyi @dimus ).

As it seems that the resolver service appears to not take action on this as of today for stated reasons (e.g., NCBI is not using proper naming conventions),  we'll have to figure out some other strategy from preventing NCBI subspecies (or virus strains) matches from globalnames from populating taxon links in GloBI.

Hoping to work on this soon. 

Thanks again for reporting this important issue and let me know if you have any ideas/ comments/ questions.

**Comment by dimus** [comment]
@pwrose yes, sorry about that. I would say the only way I can imagine to deal with this on the resolver.globalnames.org side would be to introduce case by case fixes for all the non-standard spelling of names. However I am afraid to go this route, as there are 30million name-strings in the database, and most of them get updated regularly.

**Comment by pwrose** [comment]
I see that the various hits in resolver.globalnames.org have the same match
score, and therefore cannot be distinguished.

As a workaround, I suggest to compare the query string with the results
from globalnames.org and then choose the one with the lowest edit distance (
https://en.wikipedia.org/wiki/Edit_distance).

**Comment by dimus** [comment]
this is a very good point. I cannot use edit distance in this case, because I do not compare strings to strings, that would be crazy expensive. Instead I parse them and compare canonical forms from both strings. Parsing results break into 4 categories "clean", "some problems", "big problems", "not parceable". "Homo sapiens" parses as "clean", "Homo sapiens subsp. Denisova" parses as "big problems", which  I can easily use for scoring. I am planning to retire resolver.globalnames.org, so I wont implement it there, but I will add this scoring rule to the next generation I am planning to release shortly: https://app.swaggerhub.com/apis-docs/dimus/gnames/1.0.0

Added an issue, hopefully will incorporate it this week.

https://github.com/gnames/gnames/issues/52

**Comment by pwrose** [comment]
@dimus perfect!

**Comment by dimus** [comment]
@pwrose, @jhpoelen I added parser quality scoring and it did seem to work:

(I might get server and/or db down so the curl might not work at a particular moment)

```bash
curl -X GET "https://verifier.globalnames.org/api/v1/verifications/Homo%20sapiens?pref_sources=4"|jq
```

```json
[
  {
    "inputId": "16f235a0-e4a3-529c-9b83-bd15fe722110",
    "input": "Homo sapiens",
    "matchType": "Exact",
    "bestResult": {
      "dataSourceId": 1,
      "dataSourceTitleShort": "Catalogue of Life",
      "curation": "Curated",
      "recordId": "1125941",
      "localId": "e3b90576561f93a8ac8b59e185b01511",
      "outlink": "http://www.catalogueoflife.org/annual-checklist/2019/details/species/id/e3b90576561f93a8ac8b59e185b01511",
      "entryDate": "2020-06-15",
      "matchedName": "Homo sapiens Linnaeus, 1758",
      "matchedCardinality": 2,
      "matchedCanonicalSimple": "Homo sapiens",
      "matchedCanonicalFull": "Homo sapiens",
      "currentRecordId": "1125941",
      "currentName": "Homo sapiens Linnaeus, 1758",
      "currentCardinality": 2,
      "currentCanonicalSimple": "Homo sapiens",
      "currentCanonicalFull": "Homo sapiens",
      "isSynonym": false,
      "classificationPath": "Animalia|Chordata|Mammalia|Primates|Hominoidea|Hominidae|Homo|Homo sapiens",
      "classificationRanks": "kingdom|phylum|class|order|superfamily|family|genus|species",
      "editDistance": 0,
      "stemEditDistance": 0,
      "matchType": "Exact"
    },
    "preferredResults": [
      {
        "dataSourceId": 4,
        "dataSourceTitleShort": "NCBI",
        "curation": "Curated",
        "recordId": "9606",
        "outlink": "https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&name=Homo%20sapiens&lvl=0&srchmode=1&keep=1&unlock",
        "entryDate": "2020-05-28",
        "matchedName": "Homo sapiens",
        "matchedCardinality": 2,
        "matchedCanonicalSimple": "Homo sapiens",
        "matchedCanonicalFull": "Homo sapiens",
        "currentRecordId": "9606",
        "currentName": "Homo sapiens",
        "currentCardinality": 2,
        "currentCanonicalSimple": "Homo sapiens",
        "currentCanonicalFull": "Homo sapiens",
        "isSynonym": false,
        "classificationPath": "|Eukaryota|Opisthokonta|Metazoa|Eumetazoa|Bilateria|Deuterostomia|Chordata|Craniata|Vertebrata|Gnathostomata|Teleostomi|Euteleostomi|Sarcopterygii|Dipnotetrapodomorpha|Tetrapoda|Amniota|Mammalia|Theria|Eutheria|Boreoeutheria|Euarchontoglires|Primates|Haplorrhini|Simiiformes|Catarrhini|Hominoidea|Hominidae|Homininae|Homo|Homo sapiens",
        "classificationRanks": "|superkingdom||kingdom||||phylum|subphylum|||||superclass||||class||||superorder|order|suborder|infraorder|parvorder|superfamily|family|subfamily|genus|species",
        "editDistance": 0,
        "stemEditDistance": 0,
        "matchType": "Exact"
      }
    ],
    "dataSourcesNum": 24,
    "curation": "Curated"
  }
]
```

**Comment by jhpoelen** [comment]
@pwrose @dimus great to see your exchange of ideas and preliminary results on the name linking challenges. 

I am hoping to use this results to scrub funky name matches from GloBI's taxon name links.

@dimus are you expecting this functionality to be available beyond the https://verifier.globalnames.org/api/v1/verifications/ endpoint? Also, is there a way I can run this locally, so I don't have to do a network roundtrip for each request?

**Comment by dimus** [comment]
@jhpoelen, I just got out https://github.com/gnames/gnverify (see binaries) that uses endpoint, so you still need network, but you can pipe gnverify in and out and build chains of progs. I would say you can start use it "carefully" and let me know if something breaks, or behaves wronly.

Currently system has 3 moving parts: gnmatcher, gnames, gnverify. After I make 'official' release of gnames (verify.globalnames.org) I am going to make an option of embedding gnmatcher inside of gnames. So you still have to run a gnames service on your machine, but that would make the system independent of networking.

github.com/gnames/gnmatcher can be pretty much standalone now, but need to generate its lookup files first time it runs. gnames still depends on database, and I think to leave it like this because it allows better flexibility with features. When features are stable I suspect gnames also can be moved to key-value stores like gnmatcher. With database I do have a bit of a problem, it does require to use collation C or C.UTF-8, which is not very user-friendly. I hope I can figure out something to remove this dependency on collation.

**Comment by dimus** [comment]
gnames and gnmatcher are resource hungry. gnmatcher eats ~1.5G of memory, and gnames depends on postgresql, which in turn needs some non-trivial amount of memory for faster execution.

**Comment by jhpoelen** [comment]
@dimus thanks for the description how your new tools can be used independently on the command-line. I'll need a little time to figure out what strategy works best for GloBI. In an ideal world, only a few changes would allow for picking up the newly introduced name matching quality metric.

**Comment by dimus** [comment]
I think I will make an option for gnmatcher be used as a library instead of a service to have lesser moving parts for local use. I am also thinking to put a dump of gnmatcher filters/levenschtein automata at opendata.globalnames.org and use that dump instead of database to initialize it. That should remove collation inconvenience, I think.

**Comment by jhpoelen** [comment]
As I was working on putting workarounds for Global Names greedy NCBI name matching in place, I found a duplicate issue https://github.com/GlobalNamesArchitecture/gni/issues/48 that I first filed in 2016 .

**Comment by dimus** [comment]
@jhpoelen, I am planning to retire resolver.globalnames.org as soon as I have all its functionality at verifier.globalnames.org, so I do not think it is reasonable to spend time on fixing it, I prefer to push the new system forward. The new system is live and can be used already, API is here:

https://app.swaggerhub.com/apis-docs/dimus/gnames/1.0.0

API is not immutable yet, so it might change slightly for a few months.

**Comment by jhpoelen** [comment]
@dimus what is your timeline to retire the service that I'd expect many people rely on?

**Comment by dimus** [comment]
It is a problem, that I do not know a good solution for. I am planning to put a warning for about a year or so, but if people wrote a script and use it without changes, they will not see the warning.

**Comment by dimus** [comment]
@deppaul, what do you think would be a good way to retire an ancient service minimizing suffering of its users?

**Comment by jhpoelen** [comment]
@dimus thanks for clarifying - a year long transition would work ok with me, but I am not sure whether other can respond that quickly, especially those zombie / hibernating projects that are in use but no longer maintained. I guess your traffic logs should give you an idea of different folks using it. Ideally, your valuable service would be built to last 10-20 years, with improvements happening on the background.

**Comment by dimus** [comment]
@debpaul, what do you think would be a good way to retire an ancient service minimizing suffering of its users?

**Comment by dimus** [comment]
I still run http://gni.globalnames.org which is 13 years old now, while http://resolver.globalnames.org is 10 years old. But I did not touch them for ages and I think it is about time for them to retire.

**Comment by jhpoelen** [comment]
@dimus Very cool to see that http://gni.globalnames.org and http://resolver.globalnames.org have grown up to be teenagers. To me, this inspires confidence in your ability to create stable services and software programs.

I think there's great benefit to stable services and software programs. For instance, I am enjoying a Unix program called ```cat``` (https://en.wikipedia.org/wiki/Cat_%28Unix%29) pretty much daily, and ```cat``` has been around for almost 50 years ( source: https://en.wikipedia.org/wiki/Cat_%28Unix%29 accessed at 2021-01-14 ). I wonder whether they ever considered to retire ```cat``` simply because it was too old ; )

I do have to say that software maintenance is an important and laborious task that is most commonly overlooked or excluded when new services or software libraries are first introduced. Also, the task software maintenance is often taken for granted by those who use the services: many "users" don't see the significant amount of work it takes to keep a service up and running, respond to bug reports, and implement fixes.  

So, I can empathize with your desire to stop maintaining current systems and focus on your latest greatest project that includes many of the knowledge you've gained over the last decade. As I am typing this, I am also reflecting on my own practices in developing and maintaining software and I estimate that most of my time is spent on maintaining software and services, this issue (https://github.com/globalbioticinteractions/globalbioticinteractions/issues/569) being a good example of the amount of time and effort it takes to address a seemingly small issue in linking NCBI taxa to the name 'Homo sapiens'.

On a related note: I am actively working on introducing a workaround for this and other NCBI / resolver.globalnames.org name matching challenges and hoping to introduce some fixes sooner rather than later.

Thanks for being patient and for sharing your thoughts.

**Comment by dimus** [comment]
There is also 'fmt' that on all 'normal' Linux distros still is not aware about UTF-8 existence ;)
I recall busybox fix it for their implementation though

**Comment by dimus** [comment]
I stopped investing time into `gni` and `resolver` because both of them lack a crucial feature for me: speed. That forced me to go first to Scala than to Go to fix the throughput issue. And the consequence is that instead of maintenance, there is a need to rewrite things from scratch. And supporting new features on all 4 systems would require me to be about 8 times more productive than I am now ;)

I continue to update libraries that people depend on, but maintaining big apps is a bit too much. Well, at least with `resolver` data is updated regularly so far...

**Comment by jhpoelen** [comment]
@dimus I can see how you'd want to support the (new) tools that you use yourself and help support your use-cases (e.g., indexing all of BHL!). And, I can imagine that the previous tools you've been carefully maintained over the last decade are good enough for most use cases. 

Your tools are developing rapidly and I hope I can keep up: the more time I have to spend on maintenance due to third-party service deprecation, the less time I can spend on fixing issues or adding features myself.

Thanks again for providing such useful name resolving tools and please let me know about your timelines, so I can plan accordingly.

**Comment by debpaul** [comment]
Hi @dimus I see this advice https://nordicapis.com/how-to-smartly-sunset-and-deprecate-apis/
TLDR
1. Sunset first (giving your users a time to prepare for this loss of service).
  - using [Sunset HTTP Header Field](https://tools.ietf.org/html/draft-wilde-sunset-header-11) “This specification defines the Sunset HTTP response header field, which indicates that a URI is likely to become unresponsive at a specified point in the future. It also defines a sunset link relation type that allows linking to resources providing information about an upcoming resource or service sunset.”
2. Deprecate 
  - using [Deprecation HTTP Header](https://tools.ietf.org/html/draft-dalal-deprecation-header-00#section-5): “The HTTP Deprecation response header can be used to signal to consumers of a URI-identified resource that the use of the resource has been deprecated. Additionally, the deprecation link relation can be used to link to a resource that provides additional context for the deprecation, and possibly ways in which clients can find a replacement for the deprecated resource.”

Question: Do you have any way (channels, IPs?, listserves) for contacting your users?

And another resource that shares best practices for turning of an API: see this https://swagger.io/blog/api-strategy/best-practices-for-deprecating-apis/

**Comment by dimus** [comment]
:+1: thanks @debpaul!

**Comment by jhpoelen** [comment]
@pwrose @dimus after some significant amount of work, I managed to remove the suspicious name matching results that GloBI obtained via http://resolver.globalnames.org . This includes the match with the extinct human species, but also a whole slew of virus species / strains. The procedure taken to make this happen is described in:

Poelen, Jorrit H. (2021). Global Biotic Interactions: Taxon Graph Patches (Version 0.6) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.4451462

and the resulting GloBI taxon graph can be found at

Poelen, Jorrit H. (2021). Global Biotic Interactions: Taxon Graph (Version 0.3.28) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.4451472

Collaborators at the Encyclopedia of Life @KatjaSchulz et al. provide a quite useful resource to help tease out these small issues with a giant impact by publishing a refutation dataset containing descriptions of interactions that they refuted (e.g., a virus being a host, or a mammal being pollinated by a bee). 

I was hoping to motivate myself to write a blog post about the neat interplay of reviews and automated integration processes - For instance, @pwrose 's observation about the incorrectly matched human subspecies led to a cascading effect of updates and changes spanning various different infrastructures are datasets. All this resulting in a stepwise improvement of our collective digital infrastructure. 

In short - thanks for reporting this issue and please holler if you have additional comments / questions!

PS Note that a new "stable" release of the interactions.tsv.gz is upcoming. However, the snapshot version (see https://www.globalbioticinteractions.org/data) already includes the desired updates.

**Comment by pwrose** [comment]
I noticed that https://zenodo.org/record/3950590/files/interactions.tsv.gz
lists
two targetTaxonIds for human, e.g.,  NCBI:741158 | NCBI:9606.

Would it be possible to remove NCBI:741158?

**Comment by jhpoelen** [comment]
@pwrose I've updated the Zenodo publication with a recent snapshot of the GloBI interpreted data products at:

 Poelen, Jorrit H. (2021). Global Biotic Interactions: Interpreted Data Products (Version 0.2) [Data set]. Zenodo. http://doi.org/10.5281/zenodo.4460654 .

I've re-confirmed that ```NCBI:741158``` no longer occurs in the data:

```shell
$ curl "https://zenodo.org/record/4460654/files/interactions.tsv.gz" | gunzip | grep "NCBI:741158" | wc -l
0
```

whereas over 80k interactions involving Homo sapiens (humans, NCBI:9606) were present:

```shell
$ curl "https://zenodo.org/record/4460654/files/interactions.tsv.gz" | gunzip | grep "NCBI:9606[^0-9]*" | wc -l
80436
```

Thanks for your patience: in my mind, it is these detailed conversations that make our collective digital data a little better over time . 

Please don't hesitate to open a new issue if something else comes up.

**Comment by pwrose** [comment]
I have trouble reading this file into a Pandas dataframe. It complains
about line 3626856:

ParserError: Error tokenizing data. C error: Expected 84 fields in
line 3626856, saw 119
<https://zenodo.org/record/4460654/files/interactions.tsv.gz'%3C/span%3E%3Cspan>

**Comment by jhpoelen** [comment]
@pwrose thanks for sharing your parsing error! Please use https://github.com/globalbioticinteractions/globalbioticinteractions/issues/594 for a discussion on this parsing issues that is unrelated to NCBI taxon ids (this issue).
